### ISSUE TITLE
The default mapreduce_impl doesn't work for non-Bool ==

### ISSUE THREAD

**Issue by ararslan** [JuliaLang/julia#31390]
See https://github.com/JuliaLang/julia/blob/v1.1.0/base/reduce.jl#L462. If `v == v` produces something other than `Bool`, this fails, producing errors such as that seen in https://github.com/JuliaOpt/Convex.jl/issues/275. In the case of Convex, `==` produces an `EqConstraint`.

Is the right solution here for packages to specialize `mapreduce_impl` on arrays with a particular eltype so that this logic doesn't get hit, or is there more generic solution that can replace the `v == v` check?

**Comment by mbauman** [comment]
We should probably find a better solution — the same thing happens for `missing`.

```julia
julia> maximum(Any[missing;1:15])
ERROR: TypeError: non-boolean (Missing) used in boolean context
```

Maybe just `(v == v) === true`?

**Comment by ararslan** [comment]
Actually, master seems to handle this just fine:
```
julia> maximum([missing; 1:15])
missing
```
The `v == v` check has actually been removed as of https://github.com/JuliaLang/julia/pull/30320, which inadvertently fixed this issue, both for `missing` and for Convex.

**Comment by ararslan** [comment]
I guess I'll close this, as there's nothing to be done for master. Whether this is fixed on 1.1 will be a matter of whether the change gets backported to 1.1.1.
